includes/class-sensei-course.php

```diff
@@ -116,13 +124,34 @@ public function __construct () {
         // flush rewrite rules when saving a course
         add_action('save_post', array( 'Sensei_Course', 'flush_rewrite_rules' ) );
 
+	add_action( 'template_redirect', array( __CLASS__, 'shortcode_redirect_course_login' ), 10, 1 );
+
 		// Allow course archive to be setup as the home page
 		if ( (int) get_option( 'page_on_front' ) > 0 ) {
 			add_action( 'pre_get_posts', array( $this, 'allow_course_archive_on_front_page' ) );
 		}
 	} // End __construct()
 
 	/**
+	 * If not logged in, loops through the shortcodes_auth and if the current post contains one of them
+	 * the user will be redirected to wp login
+	 *
+	 * @since 1.9.10
+	 */
+	public static function shortcode_redirect_course_login() {
+		if ( ! is_user_logged_in() ) {
+			$post_content = get_post( get_the_ID() )->post_content;
+
+			foreach ( self::$shortcodes_auth as $shortcode ) {
+				if ( has_shortcode( $post_content, $shortcode ) ) {
+					wp_redirect( wp_login_url( get_permalink() ) );
```

Comment:
To keep consistency with the previous solution, I believe it's better to re-use `sensei_get_current_page_url` than `get_permalink`.